doc/gc.rst

```diff
@@ -75,6 +75,165 @@ JavaScript's garbage collector is used for the `JavaScript and NodeJS
+syntax where the type is a `ref T`) also stores a reference counter.
+
+When the cell is allocated, the counter starts at `1` (because the allocation
+create a single `ref` handle), and is incremented when a new `ref` to the cell
+is created and decremented when a `ref` to it is destroyed.
+
+Once the counter reaches zero, the cell is immediately destroyed (via calling
+the attached destructor, if available) and the underlying memory returned to
+the allocator.
+
+In |NimSkull|, the counting is implemented via the lifetime-tracking-hook
+mechanism. When copying a `ref` (i.e. its `=copy` hook is invoked), the
+`ref` value itself (which is a pointer) is copied and the referenced cell's
+refcounter is incremented by one. Moving a `ref` only copies the pointer
+value, and destroying a `ref` (i.e. the `=destroy` hook of it is invoked)
+decrements the refcounter of the referenced cell.
+
+The `=destroy` hook for a `ref` checks the refcounter of the cell, and if it
+just reached zero, the static or virtual destructor of the cell is invoked and
+the underlying memory freed.
+
+As `ref`s use the lifetime-tracking hooks, they're subject to the same
+optimizations (i.e. cursor inference, turning copies into moves) as all other
+types with lifetime-tracking hooks.
+
+ORC
+===
+
+The problem with plain reference counting is that it is not able to handle
+reference cycles. A reference cycle exists when a cell either directly or
+indirectly stores a reference (i.e. `ref`) to itself. Example:
+
+.. code-block:: nim
+  type A = object
+    x: ref A
+
+  var a = A(x: nil) # okay; no cycle exists and the refcounter is 1
+  a.x = a # a cycle is introduced! the refcounter is 2
+
+  # once `a` gets destroyed, the refcounter becomes 1, but it's not possible
+  # for it to reach 0 from there
+  a.x = nil # explictly breaking the cycle would work
+
+  type
+    B = object
+      c: ref C
+    C = object
+      b: ref B
+
+  var b = B()
+  b.c = C() # okay; no cycle exists
+  b.c.b = b # an indirect cycle is introduced!
+
+This is where ORC becomes relevant. ORC is automatic reference-counting with a
+run-time cycle collector (the letter 'O' in ORC is meant to represent a cycle).
+
+For easier visulization, it makes sense to view a managed heap cell as a *node*,
+and a `ref` as an *edge* in a directed cyclic graph. The cycle collector is
+responsible for freeing cells only referenced as part of references cycles.
+
+The algorithm
+-------------
+
+The cycle colletor used for ORC is based on "trial deletion". As the first
+step, all edges reachable from potential *cycle roots* are temporarily
+deleted. The second step restores the outgoing edges for nodes (and nodes
+reachable from them) that still have incoming edges after the temporary
+deletion. Once done, all nodes that have no incoming edges are known to only
+be alive because of a reference cycle, and can now be freed.
+
+The implementation
+------------------
+
```

Comment:
```suggestion

Collection is implemented with a 2-pass graph colouring algorithm for detection,
and a 2-step memory collection process.

```

This intro might help setup the mental model described in the paragraphs that follow.

Reply:
Yep, that's better. Thanks.